### PR TITLE
[RLlib] DreamerV3: Catalog enhancements 04  - LSTM default models

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3: Catalog enhancements 04  - LSTM default models for tf and torch
* Support multiple LSTM layers, flexible internal state, use_bias, and batch/time major.
* Test cases compare torch w/ tf models.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/models/torch/base.py [hunk 1]
```diff
@@ -60,15 +60,11 @@ def _forward(self, inputs, **kwargs):
 
     """
 
-    def __init__(self, config: ModelConfig, skip_nn_module_init=False):
+    def __init__(self, config: ModelConfig):
```
Comment (sven1977) [inline rllib/core/models/torch/base.py:63]: This seems superflous. Wasn't implemented anyways. I think we had already fixed the class hierarchy, such that this wou;dnt be necessary anymore.

File: rllib/utils/test_utils.py [hunk 2]
```diff
@@ -1210,11 +1210,16 @@ def add(self, framework: str = "torch") -> Any:
         if isinstance(model.input_specs, SpecDict):
             inputs = {}
             for key, spec in model.input_specs.items():
-                key = key[0]
+                dict_ = inputs
```
Comment (sven1977) [inline rllib/utils/test_utils.py:1213]: This now also works for NestedDicts.

File: rllib/core/models/tf/encoder.py [hunk 3]
```diff
@@ -27,49 +30,17 @@
 _, tf, _ = try_import_tf()
 
 
-class TfMLPEncoder(Encoder, TfModel):
```
Comment (sven1977) [inline rllib/core/models/tf/encoder.py:30]: Just re-sorted this class.

File: rllib/core/models/configs.py [hunk 4]
```diff
@@ -610,17 +610,59 @@ def build(self, framework: str = "torch") -> Encoder:
...
+        # LSTM(32, 64, bias=False)
+
+        # Resulting shape of the internal states (c- and h-states):
+        # (1, B, 64) for each c- and h-states.
 
     Attributes:
-        hidden_dim: The size of the hidden layer.
-        num_layers: The number of LSTM layers.
-        batch_first: Wether the input is batch first or not.
+        input_dims: The
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/configs.py:659]: Something is missing here: The ... ? 🙂

File: rllib/core/models/tests/test_recurrent_encoders.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                f"hidden_dim: {hidden_dim}\n"
+                f"output_activation: {output_activation}\n"
+                f"output_dims: {output_dims}\n"
+                f"use_bias: {use_bias}\n"
+            )
+
+            config = LSTMEncoderConfig(
+                input_dims=inputs_dims,
+                num_lstm_layers=num_lstm_layers,
+                hidden_dim=hidden_dim,
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tests/test_recurrent_encoders.py:51]: Not that I think that it's needed, but is there a reason why we only offer a single hidden dim as opposed to making this more flexible with removing `num_lstm_layers` and making `hidden_dim` `hidden_dims`?

Comment (ArturNiederfahrenhorst) [reply rllib/core/models/tests/test_recurrent_encoders.py:51]: Is it "just" because it makes things ugly and complicated and is not really needed?

File: rllib/core/models/tf/encoder.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/models/tf/encoder.py:30]: Cool. Since the diff becomes quite weird with this, I'll only look at the LSTM code.

File: rllib/core/models/tf/encoder.py [hunk 6]
```diff
@@ -135,13 +105,145 @@ def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
...
+
+    @override(Model)
+    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+        out = tf.cast(inputs[SampleBatch.OBS], tf.float32)
+
+        # States are batch-first when coming in. Make them layers-first.
+        states_in = tree.map_structure(
+            lambda s: tf.transpose(s, perm=[1, 0] + list(range(2, len(s.shape)))),
+            inputs[STATE_IN],
+        )
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/tf/encoder.py:234]: Why are we doing this instead of doing `perm=[1, 0, 2]`?
All `state in`s will always have a single dimension right? Otherwise specs validation would also vail.
I think this code is "too" general and can be simplified to fit this particular use case, right?

File: rllib/core/models/torch/encoder.py [hunk 7]
```diff
@@ -148,91 +158,89 @@ def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
...
-        x = x.view(-1, x.shape[-1])
-
-        return {
-            ENCODER_OUT: x,
-            STATE_OUT: tree.map_structure(lambda x: x.transpose(0, 1), states_o),
-        }
+        out = inputs[SampleBatch.OBS].float()
 
+        # States are batch-first when coming in. Make them layers-first.
+        states_in = tree.map_structure(lambda s: s.transpose(0, 1), inputs[STATE_IN])
```
Comment (ArturNiederfahrenhorst) [inline rllib/core/models/torch/encoder.py:235]: I think the torch code looks cleaner here. Can me match tf to this?

File: rllib/utils/test_utils.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/test_utils.py:1213]: Cool!

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Awesome PR, thanks!

File: rllib/core/models/torch/encoder.py (hunk 7 above)
Comment (sven1977) [reply rllib/core/models/torch/encoder.py:235]: done, removed the list(range(2, .. ))) stuff for TF, as it's probably not needed (state tensors are always 1D).

File: rllib/core/models/tf/encoder.py (hunk 6 above)
Comment (sven1977) [reply rllib/core/models/tf/encoder.py:234]: done

File: rllib/core/models/tests/test_recurrent_encoders.py (hunk 5 above)
Comment (sven1977) [reply rllib/core/models/tests/test_recurrent_encoders.py:51]: Good question. Yeah, I did this for simplification reasons. We might want to generalize this in the future, but I'll leave it as-is for now. I'll add a TODO.

Comment (sven1977) [reply rllib/core/models/tests/test_recurrent_encoders.py:51]: Actually, no, torch only allows a single hidden dim value for all sub-layers within a torch.nn.LSTM. I'll just leave it at that.

File: rllib/core/models/configs.py (hunk 4 above)
Comment (sven1977) [reply rllib/core/models/configs.py:659]: great catch, fixed!
